profiling/src/collections/string_table/mod.rs

```diff
@@ -0,0 +1,279 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+#[allow(unused)]
+pub mod wordpress_test_data;
+
+use crate::collections::identifiable::{Id, StringId};
+use crate::iter::{IntoLendingIterator, LendingIterator};
+use datadog_alloc::{AllocError, Allocator, ChainAllocator, VirtualAllocator};
+use std::alloc::Layout;
+
+/// A trait that indicates an allocator is arena allocator, meaning it doesn't
+/// deallocate individual items, but deallocates their memory as a  group when
+/// the arena is dropped.
+pub trait ArenaAllocator: Allocator {
+    /// Copies the str into the arena, and returns a slice to the new str.
+    fn allocate(&self, str: &str) -> Result<&str, AllocError> {
+        let layout = Layout::for_value(str);
+        let uninit_ptr = Allocator::allocate(self, layout)?;
+
+        // Copy the bytes of the string into the allocated memory.
+        // SAFETY: this is guaranteed to not be overlapping because an
+        // allocator must not return aliasing bytes in its allocations.
+        unsafe {
+            let src = str.as_ptr();
+            let dst = uninit_ptr.as_ptr() as *mut u8;
+            let count = str.len();
+            core::ptr::copy_nonoverlapping(src, dst, count);
+        }
+
+        // SAFETY: The bytes were properly initialized, and they cannot be
+        // misaligned because they have an alignment of 1, so it is safe to
+        // create a slice of the given data and length. The lifetime matches
+        // the arena allocator's lifetime.
+        let slice: &[u8] =
+            unsafe { core::slice::from_raw_parts(uninit_ptr.as_ptr() as *const u8, str.len()) };
+
+        // SAFETY: Since the bytes were copied from a valid str without
+        // slicing, the bytes must also be utf-8.
+        Ok(unsafe { core::str::from_utf8_unchecked(slice) })
+    }
+}
+
+impl<A: Allocator + Clone> ArenaAllocator for ChainAllocator<A> {}
+
+type Hasher = core::hash::BuildHasherDefault<rustc_hash::FxHasher>;
+type HashSet<K> = indexmap::IndexSet<K, Hasher>;
+
+/// Holds unique strings and provides [StringId]s that correspond to the order
+/// that the strings were inserted.
+pub struct StringTable {
+    /// The bytes of each string stored in `strings` are allocated here.
+    bytes: ChainAllocator<VirtualAllocator>,
+
+    /// The ordered hash set of unique strings. The order becomes the StringId.
+    /// The static lifetime is a lie, it is tied to the `bytes`, which is only
+    /// moved if the string table is moved e.g.
+    /// [StringTable::into_lending_iterator].
+    /// References to the underlying strings should generally not be handed,
+    /// but if they are, they should be bound to the string table's lifetime
+    /// or the lending iterator's lifetime.
+    strings: HashSet<&'static str>,
```

Comment:
Should we take a lifetime paramater / is there way to tie this to the lifetime of the underlying object?

Reply:
Yes, there is a way to do it using `ouroboros`. The code to do it is quite ugly, because they have to restrict all kinds of things to make it safe. Under the hood, it's basically doing the same trick.